### PR TITLE
[Refactor] Enforce attaching storage scope to PointerType

### PR DESCRIPTION
**Pull request by masahi**
Right now, storage scope information is scattered across different places. This PR proposes to make `storage_scope` in PointerType, introduced in https://github.com/apache/tvm/pull/8017, be the single source of truth. This enables each `Allocate`, `Store`, `Load` nodes to directly access its storage scope through its associated buffer variable, rather than via `attr::stroage_scope` + a map between `Var` and storage scope as currently done today (for example, https://github.com/apache/tvm/blob/813136401a11a49d6c15e6013c34dd822a5c4ff6/src/tir/transforms/thread_storage_sync.cc#L341). 

The main changes are:
* In `schedule_ops.cc`, attach storage scope information to `ProduceRealizeNode`, rather than creating `AttrStmt` with `tir::attr::realize_scope` scope.
* In `schedule_postproc_to_primfunc.cc`, always create a new `Buffer` using storage scope information in `ProduceRealizeNode`. This ensures that `PointerType` associated with the new buffer has always storage scope information attached.
* In transform passes and codegen classes, try to use storage scope information in `AllocateNode` directly, and remove `attr::storage_scope` and `attr::realize_scope` when they are no longer needed. 

Building on this PR, I'll send the following changes later:
* Introduce a new storage scope of dynamic shared memory
* Remove remaining use of `attr::storage_scope`
* Remove storage scope in `Buffer` class https://github.com/apache/tvm/blob/2cca934aad1635e3a83b712958ea83ff65704316/include/tvm/tir/buffer.h#L70-L71

### CODE REVIEW

**Comment by masahi** [comment]
The CI job was killed during build twice for unclear reason.

**Comment by tqchen** [comment]
Thanks @masahi . seems there are still testcases that we need to fix

**Comment by masahi** [comment]
@tqchen I've hit an issue in `lower_thread_allreduce.cc`. There are multiple uses of `AttrStmt` where a storage scope of a buffer variable is assigned after some complicated logic, like this:
https://github.com/apache/tvm/blob/ae58f2c387de9944d241a083ce9a0dd4c9ae613d/src/tir/transforms/lower_thread_allreduce.cc#L94

So I tried to replace this line by something like `SetStorageScope(repl->buffer_var, "shared")` but this doesn't work because I cannot get a mutable pointer to `PointerType` via `buffer_var->type_annotation.as<PointerTypeNode>()`. 

Creating a new variable also doesn't work because the old `buffer_var` is [shared ](https://github.com/apache/tvm/blob/ae58f2c387de9944d241a083ce9a0dd4c9ae613d/src/tir/transforms/lower_thread_allreduce.cc#L354-L359) with other stmt nodes, so updating the `buffer_var` of `AllocateNode` results in the use of undefined variable in other nodes.

What should we do about this?

**Comment by tqchen** [comment]
In this particular case, we might need to create a new buffer var and remap the use of the old buffer vars to new one

File: Jenkinsfile [hunk 1]
```diff
@@ -282,7 +282,6 @@ stage('Unit Test') {
         timeout(time: max_time, unit: 'MINUTES') {
           sh "${docker_run} ${ci_arm} ./tests/scripts/task_ci_setup.sh"
           sh "${docker_run} ${ci_arm} ./tests/scripts/task_python_unittest.sh"
-          sh "${docker_run} ${ci_arm} ./tests/scripts/task_python_arm_compute_library.sh"
```
Comment (tqchen) [inline Jenkinsfile:285]: should not be part of the PR

File: src/target/llvm/codegen_llvm.cc [hunk 2]
```diff
@@ -498,11 +498,12 @@ void CodeGenLLVM::AddAliasInfo(llvm::Instruction* inst, const VarNode* buffer, P
 void CodeGenLLVM::GetAlignment(DataType t, const VarNode* buf_var, const PrimExpr& index,
                                int* p_alignment, int* p_native_bits) {
   int max_align_bits = t.bits();
-  auto it = alloc_storage_info_.find(buf_var);
-  if (it != alloc_storage_info_.end()) {
-    const StorageInfo& info = it->second;
```
Comment (tqchen) [inline src/target/llvm/codegen_llvm.cc:503]: Let us keep the original data structure info for now, in case we need to add more information later

File: src/tir/ir/buffer.cc [hunk 3]
```diff
@@ -45,12 +45,27 @@ Array<PrimExpr> SimplifyArray(arith::Analyzer* ana, Array<PrimExpr> array) {
...
 }
 
+String GetStorageScope(Var buffer_var) {
+  auto type = buffer_var->type_annotation;
+  const auto* ptr_type = type.as<PointerTypeNode>();
+  ICHECK(ptr_type) << "The provided variable is not of pointer type";
+  return ptr_type->storage_scope;
+}
+
+Var UpdateStorageScope(Var buffer_var, String storage_scope) {
```
Comment (tqchen) [inline src/tir/ir/buffer.cc:62]: Move this to the  caller side since this is a quick convenient function. Consider rename to WithStorageScope

File: src/tir/ir/buffer.cc [hunk 4]
```diff
@@ -45,12 +45,27 @@ Array<PrimExpr> SimplifyArray(arith::Analyzer* ana, Array<PrimExpr> array) {
...
-Buffer decl_buffer(Array<PrimExpr> shape, DataType dtype, String name, Span span) {
+Buffer decl_buffer(Array<PrimExpr> shape, DataType dtype, String name, String storage_scope,
+                   Span span) {
   DataType storage_dtype = (dtype == DataType::Bool() ? DataType::Int(8) : dtype);
-  return Buffer(Var(name, PointerType(PrimType(storage_dtype)), span), dtype, shape,
+  return Buffer(Var(name, PointerType(PrimType(storage_dtype), storage_scope), span), dtype, shape,
                 Array<PrimExpr>(), PrimExpr(), name, "", 0, 0, kDefault, span);
 }
 
+String GetStorageScope(Var buffer_var) {
```
Comment (tqchen) [inline src/tir/ir/buffer.cc:55]: Consider to the caller side since directly exposing the function is more informative.

GetPtrStorageScope

File: src/target/llvm/codegen_llvm.h [hunk 5]
```diff
@@ -161,13 +161,6 @@ class CodeGenLLVM : public ExprFunctor<llvm::Value*(const PrimExpr&)>,
   void VisitStmt_(const EvaluateNode* op) override;
 
  protected:
-  /*! \brief The storage information */
-  struct StorageInfo {
-    /*! \brief The storage scope */
-    runtime::StorageScope scope;
-    /*! \brief The alignment of allocation */
```
Comment (tqchen) [inline src/target/llvm/codegen_llvm.h:168]: Let us keep the original data structure and leave alignment field here for now, in case we need some update in the future

File: src/tir/ir/buffer.cc (hunk 3 above)
Comment (masahi) [reply src/tir/ir/buffer.cc:62]: Done

File: src/tir/ir/buffer.cc (hunk 4 above)
Comment (masahi) [reply src/tir/ir/buffer.cc:55]: Since this function is used a lot in `src/tir/transforms` and `src/tir/target`, I moved it to `src/tir/transforms/ir_utils.h`.

**Comment by masahi** [comment]
CI blocked by https://github.com/apache/tvm/pull/8400

File: src/tir/ir/stmt.cc [hunk 6]
```diff
@@ -61,6 +61,16 @@ TVM_STATIC_IR_FUNCTOR(ReprPrinter, vtable)
...
+  if (attr_key == attr::storage_scope) {
+    const VarNode* buf = node.as<VarNode>();
+    ICHECK(buf);
+    const auto* ptr_type = buf->type_annotation.as<PointerTypeNode>();
+    ICHECK(ptr_type) << "The provided variable is not of pointer type";
+    auto attr_scope = value.as<StringImmNode>()->value;
+    ICHECK(attr_scope == ptr_type->storage_scope)
+        << "Storage scopes attached to AttrStmt and buffer var are different. " << attr_scope
+        << ", " << ptr_type->storage_scope;
+  }
```
Comment (masahi) [inline src/tir/ir/stmt.cc:73]: This is the invariant that enforces the same storage scope be attached to AttrStmt and buffer variables.
Later I'll send a follow up PR to remove `attr::storage_scope` completely. 

cc @tqchen @junrushao1994 @csullivan

**Comment by kparzysz-quic** [comment]
Should we remove `scope` from `Buffer`?  The scope is now available from the type annotation of the `data` member in `BufferNode`.

**Comment by masahi** [comment]
> Should we remove `scope` from `Buffer`? The scope is now available from the type annotation of the `data` member in `BufferNode`.

Yes this is one of my follow-up items. Apparently `scope` in `Buffer` is only used by TensorIR related code.

File: src/tir/transforms/update_pointer_storage_scope.h [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+#include <tvm/tir/expr.h>
+#include <tvm/tir/op.h>
+#include <tvm/tir/stmt_functor.h>
+
+#include <unordered_map>
+
+namespace tvm {
+namespace tir {
+
+class UpdatePointerStorageScope : public StmtExprMutator {
```
Comment (masahi) [inline src/tir/transforms/update_pointer_storage_scope.h:36]: This is used by passes that require modifying storage scopes after buffer creation. Currently  used by `lower_thread_allreduce.cc` and `lower_warp_memory.cc`.

File: python/tvm/script/special_stmt.py [hunk 8]
```diff
@@ -491,6 +491,22 @@ def var(dtype, span):
         super().__init__(var, def_symbol=True)
 
 
+@register
+class BufferVarDef(SpecialStmt):
```
Comment (masahi) [inline python/tvm/script/special_stmt.py:495]: cc @spectrometerHBH for tvmscript changes. This is for distinguishing a normal variable from a buffer variable. The latter one requires storage scope. See the change in `test_tvmscript_roundtrip.py`. The tvm script printer was also updated.

**Comment by masahi** [comment]
All tests passed, ready for review.

File: src/tir/transforms/update_pointer_storage_scope.cc [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+PrimExpr UpdatePointerStorageScope::VisitExpr_(const LoadNode* op) {
+  auto remapped = StmtExprMutator::VisitExpr(op->buffer_var);
+  return Load(op->dtype, Downcast<Var>(remapped), StmtExprMutator::VisitExpr(op->index),
+              StmtExprMutator::VisitExpr(op->predicate));
+}
+
+Stmt UpdatePointerStorageScope::VisitStmt_(const AttrStmtNode* op) {
+  using runtime::StorageScope;
```
Comment (kparzysz-quic) [inline src/tir/transforms/update_pointer_storage_scope.cc:68]: You never use this name in this function.

File: src/runtime/thread_storage_scope.h [hunk 10]
```diff
@@ -118,7 +118,9 @@ struct StorageScope {
    */
   static StorageScope Create(const std::string& s) {
     StorageScope r;
-    if (s.compare(0, 6, "global") == 0) {
+    if (s == "") {
```
Comment (kparzysz-quic) [inline src/runtime/thread_storage_scope.h:121]: I suggest using `s.empty()`.

File: src/tir/transforms/update_pointer_storage_scope.cc [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+Var WithStorageScope(const VarNode* buffer_var, String storage_scope) {
+  auto* ptr_type = buffer_var->type_annotation.as<PointerTypeNode>();
+  ICHECK(ptr_type) << "The provided variable is not of pointer type";
+  return Var(buffer_var->name_hint, PointerType(ptr_type->element_type, storage_scope),
+             buffer_var->span);
+}
+
+UpdatePointerStorageScope::UpdatePointerStorageScope(
+    const std::unordered_map<const VarNode*, String>& new_storage_scopes) {
+  for (auto kv : new_storage_scopes) {
```
Comment (kparzysz-quic) [inline src/tir/transforms/update_pointer_storage_scope.cc:48]: Use `auto&`.

File: src/te/operation/cross_thread_reduction.cc [hunk 12]
```diff
@@ -177,7 +178,8 @@ Stmt MakeCrossThreadReduction(const ComputeOpNode* self, const Stage& stage,
   std::vector<Var> res_handles(size);
   for (size_t idx = 0; idx < size; ++idx) {
     DataType dtype = reduces[idx]->dtype;
-    res_handles[idx] = Var("reduce_temp" + std::to_string(idx), PointerType(PrimType(dtype)));
+    res_handles[idx] =
```
Comment (vinx13) [inline src/te/operation/cross_thread_reduction.cc:181]: shall we remove `storage_scope` attr here https://github.com/apache/tvm/blob/f459dbfd3a09e5bcba25fd6838cf878efb961fa7/src/te/operation/cross_thread_reduction.cc#L229
https://github.com/apache/tvm/blob/f459dbfd3a09e5bcba25fd6838cf878efb961fa7/src/te/operation/cross_thread_reduction.cc#L234

File: src/tir/transforms/lower_thread_allreduce.cc [hunk 13]
```diff
@@ -33,10 +33,33 @@
 
 #include "../../runtime/thread_storage_scope.h"
 #include "ir_utils.h"
+#include "update_pointer_storage_scope.h"
 
 namespace tvm {
 namespace tir {
 
+class UpdatePointerStorageScopeAllReduce final : public UpdatePointerStorageScope {
```
Comment (vinx13) [inline src/tir/transforms/lower_thread_allreduce.cc:41]: Can we merge this pass with `UpdatePointerStorageScope` by adding check for `attr::volatile_scope` in `VisitStmt_(const AttrStmtNode*)`

Comment (masahi) [reply src/tir/transforms/lower_thread_allreduce.cc:41]: This `attr::volatile_scope` is created inside this `AllocateNode` visitor. A visitor for `AttrStmtNode` don't get to see this.

File: src/te/operation/cross_thread_reduction.cc (hunk 12 above)
Comment (masahi) [reply src/te/operation/cross_thread_reduction.cc:180]: Removal of `AttrStmt` creation with `attr::storage_scope` will come after this PR, where all `attr::storage_scope` usages will be removed at once. A lot of places in the code base still assume the existence of `AttrStmt` with `attr::storage_scope`, so I don't want to remove some AttrStmt while leaving others in this PR.

**Review by kparzysz-quic — APPROVED** [review]
LGTM

File: src/tir/transforms/update_pointer_storage_scope.cc [hunk 14]
```diff
@@ -0,0 +1,91 @@
+/*
```
Comment (csullivan) [inline src/tir/transforms/update_pointer_storage_scope.cc:1]: I think we need to also update `lower_device_storage_access_info.cc` to use the PointerType storage scope in this PR as well.

With a custom storage scope used in a cache_read stage, e.g. 
```
s.cache_read(data, "global.texture", [OL])
```

I see that [lower_device_storage_access_info.cc#L72](https://github.com/apache/tvm/blob/main/src/tir/transforms/lower_device_storage_access_info.cc#L72) is unhappy with the scope tag.  
```
  Check failed: (e.info.defined()) is false: Cannot find memory info of global.texture
```

**Review by csullivan — CHANGES_REQUESTED** [review]
@masahi Thanks for the great and healthy refactor!

**Comment by tqchen** [comment]
will let @vinx13 manage this PR

File: src/tir/transforms/update_pointer_storage_scope.cc (hunk 14 above)
Comment (masahi) [reply src/tir/transforms/update_pointer_storage_scope.cc:1]: Hmm interesting. `attr::storage_scope` and the PointerType should have the same storage scope, by the invariant at https://github.com/apache/tvm/blob/3e5e0efe2dfde96f9b3a162ff66524ff79b62723/src/tir/ir/stmt.cc#L64-L73. 

A possible explanation is that the value of `scope.tag.length()` at https://github.com/apache/tvm/blob/main/src/tir/transforms/lower_device_storage_access_info.cc#L70 is different between `main` and this branch. In this case, `GetMemoryInfo` assumes that a packed function `tvm.info.mem.global.texture` is defined, but I don't find it anywhere.

Do you have a repro script?

Comment (csullivan) [reply src/tir/transforms/update_pointer_storage_scope.cc:1]: Thanks @masahi. I tracked this issue down to the buffer storage scope being set to `global` and removing the scope tag. Previously this was fine as the AttrStmtNode realize_scope annotation was used during texture lowering. 

This demonstrate the value in removing the buffer storage scope in favor of consolidating around the PointerType's storage_scope. I look forward to that PR! Thanks again for this change.

**Review by csullivan — APPROVED** [review]
LGTM

**Review by vinx13 — APPROVED** [review]

**Comment by vinx13** [comment]
Thanks @masahi @kparzysz-quic @csullivan @tqchen this is merged
